prov-api/routes/explore/explore.py

```diff
@@ -316,3 +321,115 @@ async def effected_agents(
+        FastAPI background task handler which manage tasks such as cleaning up temporary files. 
+
+    Returns
+    -------
+    FileResponse
+        A FileResponse object that contains the generated word-document and is sent to the front-end. 
+
+    Raises
+    ------
+    HTTPException
+        Raised if the input depth is invalid (not within the range of 1 to 3).
+    HTTPException
+        Raised if an unsupported item subtype is provided.
+    HTTPException
+        Raised if there is an error in generating the word document.
+
+    """
+         
+    # Create the request style, here we assert where the HTTP request came from. 
+    request_style: RequestStyle = RequestStyle(
+        user_direct=roles.user, service_account=None
+    )
+
+    # Validate the depth.
+    if depth < 1 or depth > 3: 
+        raise HTTPException(status_code=400, detail="Invalid input depth provided. Input depth is only from 1 to 3")
+    
+    # do checks accordingly. 
+    await validate_node_id(node_id=node_id, item_subtype=item_subtype, request_style=request_style, config=config)
+
+    # This acting as a shared dictionary across both edge cases and in the helper function in the edge cases.
+    # Could be converted into a dataclass maybe??
+    all_filtered_responses: Dict[str, List[Node]] = {"inputs": [], "outputs": [], "model_runs": []}
+
+    if item_subtype == ItemSubType.MODEL_RUN: 
+        await generate_report_helper(starting_id=node_id, upstream_depth=depth, shared_responses=all_filtered_responses, config=config)
+
+    elif item_subtype == ItemSubType.STUDY: 
+
+        """
+        Upstream (Outputs):
+        Example: From a study, it will fetch all model runs associated with the study and their inputs, up to 3 layers upstream.
+
+        Downstream (Outputs): 
+        Example: From a study, it will fetch all relevant datasets (outputs) from the model runs linked to the study, up to 1 layer downstream.
+        """
+    
+        # Query upstream to get all linked model runs at depth 1. 
+        # This may contain more than one model run.
+        upstream_model_run_response = await explore_downstream(starting_id=node_id, depth=1, config=config)
+        assert upstream_model_run_response.graph, "Node collections not found!"
+
+        model_run_nodes: List[Node]= parse_nodes(upstream_model_run_response.graph.get('nodes', []))
+
+        # Now branch out with the model runs in here and explore them as above (depth 1-3 upstream and depth 1 downstream)
+        # The model run will be the new updated starting id.
+
+        for model_run_node in model_run_nodes:
+            if model_run_node.item_subtype == ItemSubType.MODEL_RUN:
+                model_run_id = model_run_node.id
+                # Extract the upstream and downstream entities from this node. 
+                await generate_report_helper(starting_id=model_run_id, upstream_depth=depth, shared_responses=all_filtered_responses, config=config)                
```

Comment:
could you chuck this into a function?

Reply:
I have broken this down into different sub-functions, should help with readability.